File: DFS-BFS-4-travelPath/travelPath3.cpp
#include <bits/stdc++.h>

using namespace std;

/*
티켓들을 정렬하고  첫 번째 경로를 찾으면 바로 리턴
10000번도 처리 가능

노드를 dfs로 방문하는데 방문할데가 없으면 리턴하고 스택에 추가
<생소하지만 매우 빠름>
dfs 깊은 곳에서부터 스택에 차례로 넣고  탐색할 dfs 있으면 탐색 후 가장 깊은 곳에서부터 스택에 차례로 넣음


방문하지 않고 다음 노드에 해당하면 방문처리하고 dfs 탐색
더 이상 dfs 들어갈 곳이 없다면 리턴

이전 dfs가 소멸하는 과정에 먼저 스택에 마지막 노드 추가함
다른 다음 dfs를 탐색하고
더 이상 dfs 들어갈 곳이 없다면 리턴
이것을 반복


*/

stack<string> st;

// 맨 아래에 마지막 공항이, 맨 위에 첫 공항이오도록 스택 쌓임
int DFS(vector<vector<string>> &tickets, vector<bool> &visited, string start)
{
    for (int i = 0; i < tickets.size(); ++i)
    {
        if (!visited[i])
        {
            if (tickets[i][0] == start)
            {
                visited[i] = true;
                DFS(tickets, visited, tickets[i][1]); // 알파벳 앞선 다음 도착지만 찾아나가면 무조건 모든 경로 거칠 수 있나? 아님
                st.push(tickets[i][1]);
            }
        }
    }
    return 0;
}

bool compare(vector<string> a, vector<string> b)
{
    return a[1] < b[1];
}

vector<string> solution(vector<vector<string>> tickets)
{
    vector<string> answer;
    vector<bool> visited;

    for (int i = 0; i < tickets.size(); ++i)
        visited.push_back(false);

    sort(tickets.begin(), tickets.end(), compare); // 티켓의 도착지를 기준으로 오름차순 정렬

    cout << "tickets: ";
    for (int i = 0; i < tickets.size(); i++)
    {
        cout << "{" << tickets[i][0] << ", " << tickets[i][1] << "}" << ' ';
    }
    cout << endl;

    // DFS(tickets, visited, "ICN");
    DFS(tickets, visited, "a");

    // answer.push_back("ICN");
    answer.push_back("a");
    while (!st.empty())
    {
        answer.push_back(st.top());
        st.pop();
    }

    return answer;
}

int main(void)
{
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<int> dis_airport_num(3, 10000); // 주어진 공항 수는 3개 이상 10,000개 이하
    uniform_int_distribution<int> dis_alphabet(65, 90);      // 'A' ~ 'Z'

    int airport_num = dis_airport_num(gen);
    airport_num = 10000;

    string fr = "ICN";
    vector<vector<string>> tickets;
    for (int i = 0; i < airport_num; i++)
    {
        vector<string> ticket;
        string to = "";
        ticket.push_back(fr);
        for (int j = 0; j < 3; j++)
        {
            to += char(dis_alphabet(gen));
        }
        if (fr != to) // 티켓의 출발지와 도착지가 달라지도록 처리
        {
            ticket.push_back(to);
            fr = to;
            tickets.push_back(ticket);
        }
        else
            i--;
    }

    shuffle(tickets.begin(), tickets.end(), gen); // 배열의 순서를 랜덤으로 셔플

    // tickets = {{"ICN", "SFO"}, {"ICN", "ATL"}, {"SFO", "ATL"}, {"ATL", "ICN"}, {"ATL", "SFO"}};
    tickets = {{"a", "g"}, {"g", "h"}, {"h", "i"}, {"i", "j"}, {"j", "e"}, {"e", "f"}, {"a", "w"}, {"w", "a"}};

    // cout << "tickets: ";
    // for (int i = 0; i < tickets.size(); i++)
    // {
    //     cout << "{" << tickets[i][0] << ", " << tickets[i][1] << "}" << ' ';
    // }
    // cout << endl;

    vector<string> result;
    auto start_time = chrono::steady_clock::now();

    result = solution(tickets);

    auto end_time = chrono::steady_clock::now();
    chrono::duration<double> elapsed_seconds = end_time - start_time;
    cout << "--- " << elapsed_seconds.count() << " seconds ---" << endl;
    cout << result.size() << ' ';
    cout << "result: ";
    for (int i = 0; i < result.size(); i++)
    {
        cout << result[i] << " ";
    }
    cout << endl;

    return 0;
}

File: DFS-BFS-4-travelPath/travelPath3.py
import random
import time


'''
알파벳이 앞선 다음 도착지로 이동 및 스택에 추가 
제일 빠름 
재귀 깊이 제한도 문제 없음
dfs 아닌 스택 반복문 이용

알파벳 순으로 다음꺼 찾으면 티켓츠에서 제거하고 스택에 쌓음
못찾으면 스택에서 제거하고 패쓰에 추가함
'''


def solution(tickets):
    routes = {}           # 빈 사전
    for t in tickets:
        # key = 출발지, values = 도착지 리스트
        # .get(a, []),  a 키값이 없을 시 default로 [], 그렇지 않으면 키에 맞는 밸류값을 반환
        routes[t[0]] = routes.get(t[0], []) + [t[1]]
    for r in routes:
        routes[r].sort(reverse=True)        # 도착지 리스트를 내림차순 정렬
    stack = ["ICN"]
    path = []
    while len(stack) > 0:
        top = stack[-1]
        # 다음 도착지로 연결될 티켓이 없거나 티켓을 다 사용해서 못쓰는 경우
        if top not in routes or len(routes[top]) == 0:
            # 스택에서 그 경로를 뺌, 다른 길을 찾기 위함 (해당 길이 막히면 한칸 뒤로가서 다른 길 찾아나감, 반복)
            path.append(stack.pop())
            # path에 추가함 (거꾸로 맨 도착지부터 출발지로 차례대로 path에 추가)
            # 중간 경로가 끊긴다는 건 그 지점이 마지막으로 가야한다는 걸 의미함
        else:
            stack.append(routes[top][-1])   # 알파벳 순서가 앞선 도착지를 스택에 넣고
            routes[top] = routes[top][:-1]  # 도착 리스트에서 제외함
    return path[::-1]


airport_num = random.randint(3, 10000)  # 주어진 공항 수는 3개 이상 10,000개 이하
# airport_num = 10000
tickets = []
fr = "ICN"
for _ in range(airport_num):
    to = ""
    for _ in range(3):
        to += chr(random.randint(65, 90))  # A(65) ~ Z(90)
    ticket = [fr, to]      # [출발, 도착] 티켓
    fr = to
    tickets.append(ticket)  # 티켓 리스트
random.shuffle(tickets)  # 티켓 리스트 셔플
# print(tickets)

tickets = [["ICN", "SFO"], ["ICN", "ATL"], [
    "SFO", "ATL"], ["ATL", "ICN"], ["ATL", "SFO"]]


start_time = time.time()
result = solution(tickets)
print("--- %s seconds ---" % (time.time() - start_time))
print(result)
